src/languageSupport/languages/flux/lsp/connection.ts

```diff
@@ -91,6 +111,193 @@ class LspConnectionManager {
     this._worker.postMessage(msg)
   }
 
+  _toggleSchemaSync() {
+    if (this._session.composition.synced) {
+      return this._callbackSetSession({...DEFAULT_SCHEMA})
+    }
+    this._callbackSetSession({
+      ...this._session,
+      composition: {...this._session.composition, synced: true},
+    })
+  }
+
+  _setCompositionSyncToggle() {
+    setTimeout(() => {
+      const syncIcon = document.getElementsByClassName(ICON_SYNC_ON)[0]
+      const clickableInvisibleDiv = document.getElementById(ICON_SYNC_ID)
+      if (!syncIcon || !clickableInvisibleDiv) {
+        return
+      }
+      clickableInvisibleDiv.style.top =
+        ((syncIcon as any).offsetTop || 0) + 'px'
+      clickableInvisibleDiv.style.left = (syncIcon as any).offsetLeft + 'px'
+      clickableInvisibleDiv.style.height =
+        ((syncIcon as any).offsetHeight || 0) + 'px'
+      clickableInvisibleDiv.style.width =
+        ((syncIcon as any).offsetWidth || 0) + 'px'
+
+      clickableInvisibleDiv.removeEventListener('click', () =>
+        this._toggleSchemaSync()
+      ) // may have existing
+      clickableInvisibleDiv.addEventListener('click', () =>
+        this._toggleSchemaSync()
+      )
+    }, 1000)
+  }
+
+  _setCompositionIrreversibleExit(schema: SchemaSelection) {
+    this._model.onDidChangeContent(e => {
+      const {changes} = e
+      changes.forEach(change => {
+        if (
+          change.range.startLineNumber >=
+            (schema.composition.block?.startLine || COMPOSITION_INIT_LINE) &&
+          change.range.endLineNumber <=
+            (schema.composition.block?.endLine || COMPOSITION_INIT_LINE)
+        ) {
+          // TODO: eventually, will be based on LSP response
+          this._callbackSetSession({
+            ...this._session,
+            composition: {...this._session.composition, canInit: false},
+          })
+          this._toggleSchemaSync()
+          return
+        }
+      })
+    })
+  }
+
+  _setCompositionBlockSize() {
+    const query = this._model.getValue()
+    const endLine =
+      query.split('\n').findIndex(line => line.includes(COMPOSITION_YIELD)) + 1
+    this._callbackSetSession({
+      ...this._session,
+      composition: {
+        ...this._session.composition,
+        block: {startLine: COMPOSITION_INIT_LINE, endLine},
+      },
+    })
+  }
+
+  _initComposition(schema: SchemaSelection) {
+    if (!this._session.composition.canInit) {
+      return
+    }
+
+    // msg to LSP
+    const {bucket, measurement} = schema
+    const payload = {bucket: bucket?.name}
+    if (measurement) {
+      payload['measurement'] = measurement
+    }
+    // TODO: finish LSP update first
+    // this.inject(ExecuteCommand.CompositionInit, payload)
```

Comment:
After the LSP schema composition service is online, then we call `this.inject()` with the proper payload.